### ISSUE TITLE
nonlocal type assignment error not caught

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#2900]
**Describe the bug**

```py
from typing import Callable

def patch_method() -> Callable[[], None]:
    captured = 5

    def inner() -> None:
        nonlocal captured
        reveal_type(captured)  # T: int
        captured = 'a'  # should error
        reveal_type(captured)  # T: Literal['a']

    return inner
```

**Expected behavior**

A type error should be raised on the incorrect assignment of the `captured` variable within the `inner` function.

**VS Code extension or command-line**

Command-line v1.1.212

**Additional context**

Should be noted that mypy raises an error

```
main.py:8: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
            captured = 'a'
```

**Comment by erictraut** [comment]
The behavior is correct. You haven't declared the type for the variable `captured`, so pyright allows you to assign any type to it, and it infers its type to be the union of all assigned types. This is the way pyright handles all variables with undeclared types; it's not specific to nonlocal types. Mypy, by contrast, assumes that a variable's type should be restricted to its "first" assignment (where "first" is in file order, not necessarily in runtime chronological order). This leads to a bunch of false positive errors and leads to much confusion, so most people prefer pyright's approach here.

If your intent here is to limit `captured` to only accept `int` values, then you should annotate its type as such.

**Comment by RobertCraigie** [comment]
Ah I see, I had forgotten about that behaviour.

I was originally casting instead of annotating the variable definition as a work around https://github.com/microsoft/pyright/issues/2757#issuecomment-1000429357. Would it make sense to treat a variable that is defined with a cast with the same heuristics as if it was annotated?

**Comment by erictraut** [comment]
No, a `cast` operates on an expression and overrides the evaluated type for that expression. It can be used anywhere that expressions are used. It has no relation to type declarations.

A type declaration is a clear expression by the programmer that "this symbol has this type". Once a symbol has a declared type, it's the job of the type checker to enforce that invariant.

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-declarations).
